### PR TITLE
Add a reentrancy module

### PR DESCRIPTION
**Pull request by nfurfaro**
This PR:
- adds a new module `reentrancy.sw`
- exposes `is_reentrant()` & `reentrancy_guard()` functions
- moves helper functions to `call_frames.sw`
- adds better documentation for const values 
- adds tests

PR discussion history: https://github.com/FuelLabs/sway-lib-std/pull/34

### CODE REVIEW

**Comment by nfurfaro** [comment]
I'm adding at least 1 more test and reviewing the logic more closely before requesting reviews here.

**Comment by nfurfaro** [comment]
Hmm. One failing test... it's making we question my assumptions about how this is working. As far as I can see, the `reentrancy::can_call_guarded_function`  test should be passing.

**Comment by nfurfaro** [comment]
Ok, after digging into this for a while, I have a few notes:
- all tests are failing. (2 show passing as they are expected to revert, but they're reverting for the wrong reason). I will update tests to capture this.
- all tests failing with `ContractCallError("MemoryOverflow")`
- while looking hard at this, I did find a bug in my implementation for `is_reentrant()`. 
- TL:DR: I previously had this line updating the  `call_frame_pointer` value:
`call_frame_pointer = get_previous_frame_pointer(call_frame_pointer);`
after the id comparison `if previous_contract_id == this_id {...`
which effectively was causing the function to always return `true`. (because `previous_contract_id` would then always be the same as `this_id` for the first iteration).

File: sway-lib-std/src/context/call_frames.sw [hunk 1]
```diff
@@ -5,6 +5,14 @@ library call_frames;
 
 use ::contract_id::ContractId;
 
+// Note that everything when serialized is padded to word length.
+//
+// Call Frame        :  saved registers = 8*WORD_SIZE = 8*8 = 64
+// Reserved Registers:  frame pointer   = 6*WORD_SIZE = 6*8 = 48
+
+const SAVED_REGISTERS_OFFSET = 64;
+const CALL_FRAME_OFFSET = 48;
```
Comment (adlerjohn) [inline sway-lib-std/src/context/call_frames.sw:14]:
```suggestion
const PREV_FRAME_POINTER_OFFSET = 48;
```

Comment (adlerjohn) [reply sway-lib-std/src/context/call_frames.sw]: Please update this constant's name throughout.

File: sway-lib-std/src/context/call_frames.sw [hunk 2]
```diff
@@ -49,3 +57,24 @@ pub fn second_param() -> u64 {
...
 }
+
+///////////////////////////////////////////////////////////
+//  Accessing arbitrary call frames by pointer
+///////////////////////////////////////////////////////////
+
+/// get a pointer to the previous (relative to the 'frame_pointer' param) call frame using offsets from a pointer.
+pub fn get_previous_frame_pointer(frame_pointer: u64) -> u64 {
+    // let offset = SAVED_REGISTERS_OFFSET + CALL_FRAME_OFFSET;
+    let offset = frame_pointer + 64 + 48;
```
Comment (otrho) [inline sway-lib-std/src/context/call_frames.sw:68]:
```suggestion
    let offset = SAVED_REGISTERS_OFFSET + CALL_FRAME_OFFSET;
```

File: sway-lib-std/src/reentrancy.sw [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    // Reentrancy cannot occur in an external context.
+    // If not detected by the time we get to the bottom of the call_frame stack,
+    // then no reentrancy has occured.
+    while call_frame_pointer != 0 {
+        call_frame_pointer = get_previous_frame_pointer(call_frame_pointer);
+        // get the ContractId value from the previous call frame
+        let previous_contract_id = get_contract_id_from_call_frame(call_frame_pointer);
+        if previous_contract_id == this_id {
+            reentrancy = true;
```
Comment (otrho) [inline sway-lib-std/src/reentrancy.sw:36]: You could just `return true` here and not need a `reentrancy` bool at all.

File: sway-lib-std/tests/test_artifacts/reentrancy_attacker_contract/src/main.sw [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        let id = target.value;
+        let target = abi(Target, id);
+        target.cross_function_reentrance_denied();
+    }
+
+    fn innocent_call(target: ContractId) -> bool {
+        let id = target.value;
+        let target = abi(Target, id);
+        let return_value = target.guarded_function_is_callable();
+        return_value
```
Comment (otrho) [inline sway-lib-std/tests/test_artifacts/reentrancy_attacker_contract/src/main.sw:47]:
```suggestion
        target.guarded_function_is_callable()
```

File: sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    } else {
+        let unwrapped = result.unwrap();
+        if let Sender::ContractId(v) = unwrapped {
+            ret = v;
+        } else {
+            panic(0);
+        };
+    };
+
+    ret
```
Comment (otrho) [inline sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw:21]: I know these are tests so style doesn't matter too much, but if we want to encourage a Rusty way of using Sway then I'd try to avoid mutable variables, and `unwrap()` as much as possible.

```suggestion
    match result {
        Result::Ok(Sender::ContractId(v)) => v,
        _ => panic(0),
    }
```

This is completely untested, I've edited this in a GitHub text box, so don't use it.  I have a horrible feeling that it wouldn't type check either?  Ideally `panic()` returns T which matches whatever it has to so it will type check, knowing that it actually never returns, but this sounds like too advanced behaviour for Sway.

To make it type check you could use `{ panic(0); ~ContractId::from(ZERO) }` for the failure arm, though that's annoying.

**Review by otrho — CHANGES_REQUESTED** [review]
I admit I don't quite follow the test logic, but in general it looks good!  Just a couple of minor things to fix.

File: sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw (hunk 5 above)
Comment (adlerjohn) [reply sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw:21]: I don't think that's possible yet, hence why `if let` was implemented first as an easy way to unblock `Result.unwrap()`.

Comment (otrho) [reply sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw:21]: What's not possible?  That match could also be an `if let`:
```
    if let Result::Ok(Sender::ContractId(v)) = result {
        v
    } else {
        panic(0);
        ~ContractId::from(ZERO)
    }
```
Is that pattern not supported?

File: sway-lib-std/src/context/call_frames.sw (hunk 1 above)
Comment (nfurfaro) [reply sway-lib-std/src/context/call_frames.sw]: Will do. Clicking the "commit suggestion" button makes it too easy to miss this stuff.

File: sway-lib-std/src/context/call_frames.sw (hunk 2 above)
Comment (nfurfaro) [reply sway-lib-std/src/context/call_frames.sw:68]: Now that we have this thing working with tests I'm happy to give it a clean up pass. I'll have to double check tomorrow, but the last time I tried to actually use these constant values here I got this:

// let offset = CODE_SIZE_OFFSET;
                              // ^^^^^^^^^^^^^^^^ Internal compiler error: Unknown variable in assembly generation. This should have been an error during type checking.Please file an issue on the repository and include the code that triggered this error.

If it persists I'll open an issue.

File: sway-lib-std/src/reentrancy.sw (hunk 3 above)
Comment (nfurfaro) [reply sway-lib-std/src/reentrancy.sw:36]: I will revisit this. There was a reason I chose this way, I think it was this error:

```
let previous_contract_id = get_contract_id_from_call_frame(call_frame_pointer);
33 |           if previous_contract_id == this_id {
   |  _________^
34 | |             true
35 | |             // reentrancy = true;
36 | |             // call_frame_pointer = 0;
37 | |         }
38 | |     }
   | |____^ Mismatched types.
expected: ()
found:    bool.
help: A while loop's loop body cannot implicitly return a value.Try assigning it to a mutable variable declared outside of the loop instead.
```

But I'll try with an explicit `return`.

File: sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw (hunk 5 above)
Comment (adlerjohn) [reply sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw:21]: IIRC that `match` version you wrote isn't possible yet. The `if let` version is, and is definitely a cleanup!

Comment (nfurfaro) [reply sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw:21]: I would _love_ to be using match statements, and I think style always matters. 
I have a feeling I'll have to use the workaround you suggested but I'm ok with that, it's still way cleaner.

As for the unwrap(), in this case it seemed ok, as the result is a `Result<Enum<T, T>, Error>`.
The first call to `is_err()` should handle the `Err` case ( in this case with a panic which is fine), and then I'm unwrapping the Sender enum so I can extract the inner value. 

But overall I'll use `match` over something like this anyday.

File: sway-lib-std/tests/test_artifacts/reentrancy_attacker_contract/src/main.sw (hunk 4 above)
Comment (nfurfaro) [reply sway-lib-std/tests/test_artifacts/reentrancy_attacker_contract/src/main.sw:47]: I'll try to clean the tests up, they drifted a bit while just make it work.
This is what I had originally. I remember assigning to a variable to silence a compiler warning, but I should be able to make this better.

File: sway-lib-std/src/context/call_frames.sw (hunk 2 above)
Comment (otrho) [reply sway-lib-std/src/context/call_frames.sw:68]: We have a couple of `const` tests in the E2E, though I think the compiler complains if you don't give them a type annotation.

File: sway-lib-std/src/reentrancy.sw (hunk 3 above)
Comment (otrho) [reply sway-lib-std/src/reentrancy.sw:36]: Yeah, it'd need to be explicit.

File: sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw (hunk 5 above)
Comment (nfurfaro) [reply sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw:21]: wouldn't the implicit return of the ContractId in 
```
} else {
        panic(0);
        ~ContractId::from(ZERO)
    }
```

be dead code?

Comment (otrho) [reply sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw:21]: Yep, but I don't think it will type-check without it.

Comment (adlerjohn) [reply sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw:21]: https://github.com/FuelLabs/sway/pull/964

Comment (otrho) [reply sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw:21]: Oh yeah.  So panic() should type check after all.  Cool.
```
    if let Result::Ok(Sender::ContractId(v)) = result {
        v
    } else {
        panic(0);
    }
```
But only if that `if let` pattern match will work.

Comment (nfurfaro) [reply sway-lib-std/tests/test_artifacts/reentrancy_target_contract/src/main.sw:21]: fixed.

File: sway-lib-std/src/reentrancy.sw (hunk 3 above)
Comment (nfurfaro) [reply sway-lib-std/src/reentrancy.sw:36]: So I tried to clean this up with an explicit return with no luck. Everything I tried either:
- caused tests to fail with `OutOfGas`, or
- caused tests to fail with `fatal runtime error: stack overflow`

Comment (nfurfaro) [reply sway-lib-std/src/reentrancy.sw:36]: For reference, this overflows the stack:
```
pub fn is_reentrant() -> bool {
    let mut reentrancy = false;
    let mut call_frame_pointer = frame_ptr();
    let this_id = contract_id();
    while call_frame_pointer != 0 {
        call_frame_pointer = get_previous_frame_pointer(call_frame_pointer);
        let previous_contract_id = get_contract_id_from_call_frame(call_frame_pointer);
        if previous_contract_id == this_id {
            call_frame_pointer = 0;
            return true;
        };
    }
    false
}
```
And this causes `OutOfGas` errors:
```
pub fn is_reentrant() -> bool {
    let mut reentrancy = false;
    let mut call_frame_pointer = frame_ptr();
    let this_id = contract_id();
    while call_frame_pointer != 0 {
        call_frame_pointer = get_previous_frame_pointer(call_frame_pointer);
        let previous_contract_id = get_contract_id_from_call_frame(call_frame_pointer);
        if previous_contract_id == this_id {
            return true;
        };
    }
    false
}
```

File: sway-lib-std/src/context/call_frames.sw (hunk 2 above)
Comment (nfurfaro) [reply sway-lib-std/src/context/call_frames.sw:68]: This seems to be a bug. I opened an issue here: https://github.com/FuelLabs/sway/issues/1125

File: sway-lib-std/tests/test_artifacts/reentrancy_attacker_contract/src/main.sw (hunk 4 above)
Comment (nfurfaro) [reply sway-lib-std/tests/test_artifacts/reentrancy_attacker_contract/src/main.sw]: cleaned up.

**Comment by nfurfaro** [comment]
Re: the test logic

Definitely a bit convoluted!
TL;DR;
- In all cases but 1, the SDK calls the `Attacker` contract, which calls the `Target` contract, which calls back into the `Attacker` contract, which calls back into (reenters) the `Target` contract before the first call to the `Target` contract has finished executing.
- The exception to this pattern is the `can_call_guarded_function()` test, which just demonstrates that a function guarded by the `reentrancy_guard()` function can still be successfully called (as long as the call is not reentrant).


The other tests show that we can:
- detect reentrancy and act on it in any way we want
- strictly forbid reentrant calls
- detect reentrancy across functions: 
  - Technically, our method of detecting reentrancy operates at the contract level, not at the function level. So, if `Attacker` contract, calls  `Target::foo()` , which calls back into the `Attacker` contract, which calls back into `Target` contract (but this time calls  `Attacker::bar()`), `is_reentrant()` inside of `Target::bar()` will return true. 
  - This is not ideal, but I think it's ok. This does not prevent `ContractA` from making multiple successive calls into `ContractB`, only from doing so in a reentrant manner. I can't think of a good reason to do the latter, but would love feedback on this!

cc @adlerjohn @otrho

File: sway-lib-std/src/reentrancy.sw (hunk 3 above)
Comment (otrho) [reply sway-lib-std/src/reentrancy.sw:34]: #1128

**Review by otrho — APPROVED** [review]
Are the tests still failing?

**Comment by nfurfaro** [comment]
@otrho All tests are passing in the current state.
The errors mentioned resulted when I tried to implement suggested changes (adding an explicit `return true;`).

File: sway-lib-std/tests/test_projects/reentrancy/mod.rs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        .tx_params(TxParameters::new(Some(0), Some(MAX_GAS_PER_TX), None))
+        .call()
+        .await
+        .unwrap();
+
+    assert_eq!(result.value, true);
+}
+
+#[tokio::test]
+#[should_panic(expected = "RESERV00")]
```
Comment (adlerjohn) [inline sway-lib-std/tests/test_projects/reentrancy/mod.rs:40]: This will have to change once https://github.com/FuelLabs/fuel-tx/pull/99 is percolated up

**Review by adlerjohn — APPROVED** [review]
